src/Microsoft.AspNetCore.Razor.Evolution/DirectiveTokenDescriptorComparer.cs

```diff
@@ -0,0 +1,44 @@
+﻿// Copyright (c) .NET Foundation. All rights reserved.
+// Licensed under the Apache License, Version 2.0. See License.txt in the project root for license information.
+
+using System;
+using System.Collections.Generic;
+using Microsoft.Extensions.Internal;
+
+namespace Microsoft.AspNetCore.Razor.Evolution
+{
+    internal class DirectiveTokenDescriptorComparer : IEqualityComparer<DirectiveTokenDescriptor>
+    {
+        public static readonly DirectiveTokenDescriptorComparer Default = new DirectiveTokenDescriptorComparer();
+
+        protected DirectiveTokenDescriptorComparer()
+        {
+        }
+
+        public bool Equals(DirectiveTokenDescriptor descriptorX, DirectiveTokenDescriptor descriptorY)
+        {
+            if (descriptorX == descriptorY)
+            {
+                return true;
+            }
+
+            return descriptorX != null &&
+                string.Equals(descriptorX.Value, descriptorY.Value, StringComparison.Ordinal) &&
```

Comment:
Huh?

If `descriptorX != descriptorY` and `descriptorX != null`, `descriptorY` could be anything except the same reference as `descriptoryX` -- including `null`. That is, the code currently excludes only a single non-`null` value for `descriptorY`. The code you checked in can easily null ref.

Second, two `null` references should compare equal. I was suggesting an _addition_:
``` c#
if (descriptorX == descriptorY)
{
    return true;
}

if (descriptorX == null || descriptorY == null)
{
    return false;
}

return (string.Equals(descriptorX.Name, descriptorY.Name, StringComparison.Ordinal) &&
    descriptorX.Kind == descriptorY.Kind &&
    Enumerable.SequenceEqual( ... );
```

I had a look at the other comparers in Razor. Primarily the test ones are `null`-safe; they often `Assert.NotNull()` for both parameters.